src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.ExtensionMethodSymbolComputer.cs

```diff
@@ -0,0 +1,539 @@
+                    // to the receiverTypeSymbol, then no need to proceed further.
+                    if (CheckedReceiverTypes.TryGetValue(receiverType, out var cachedResult) && !cachedResult)
+                    {
+                        continue;
+                    }
+
+                    // We haven't seen this type yet. Try to check by reducing one extension method
+                    // to the given receiver type and save the result.
+                    if (!cachedResult)
+                    {
+                        var reducedMethodSymbol = methodSymbols.First().ReduceExtensionMethod(ReceiverTypeSymbol);
+                        cachedResult = reducedMethodSymbol != null;
+                        CheckedReceiverTypes[receiverType] = cachedResult;
+                    }
+
+                    // Receiver type matches the receiver type of the extension method declaration.
+                    // We can add accessible ones to the item builder.
+                    if (cachedResult)
+                    {
+                        foreach (var methodSymbol in methodSymbols)
+                        {
+                            if (SemanticModel.IsAccessible(Position, methodSymbol))
+                            {
+                                builder.Add(methodSymbol);
+                            }
+                        }
+                    }
+                }
+
+                return builder.ToImmutable();
+            }
+
+            private MultiDictionary<ITypeSymbol, IMethodSymbol> GetPotentialMatchingSymbolsFromAssembly(
+                IAssemblySymbol assembly,
+                MultiDictionary<string, (string methodName, string receiverTypeName)> extensionMethodFilter,
+                bool internalsVisible,
+                CancellationToken cancellationToken)
+            {
+                var builder = new MultiDictionary<ITypeSymbol, IMethodSymbol>();
+
+                foreach (var (fullyQualifiedContainerName, methodInfo) in extensionMethodFilter)
+                {
+                    // First try to filter out types from already imported namespaces
+                    var indexOfLastDot = fullyQualifiedContainerName.LastIndexOf('.');
+                    var qualifiedNamespaceName = indexOfLastDot > 0 ? fullyQualifiedContainerName.Substring(0, indexOfLastDot) : string.Empty;
+
+                    if (NamespaceInScope.Contains(qualifiedNamespaceName))
+                    {
+                        continue;
+                    }
+
+                    // Container of extension method (static class in C# and Module in VB) can't be generic or nested.
+                    var containerSymbol = assembly.GetTypeByMetadataName(fullyQualifiedContainerName);
+
+                    if (containerSymbol == null
+                        || !containerSymbol.MightContainExtensionMethods
+                        || !IsAccessible(containerSymbol, internalsVisible))
```

Comment:
i have to admit i don't 100% get what's going on here (for the whole method).  comments might help.